packages/powersync_core/lib/src/sync/sync_status.dart

```diff
@@ -0,0 +1,377 @@
+        targetCount: bucket.count ?? 0,
+      );
+    }
+
+    return InternalSyncDownloadProgress(buckets);
+  }
+
+  static InternalSyncDownloadProgress ofPublic(SyncDownloadProgress public) {
+    return public._internal;
+  }
+
+  /// Sums the total target and completed operations for all buckets up until
+  /// the given [priority] (inclusive).
+  ProgressWithOperations untilPriority(BucketPriority priority) {
+    final (total, downloaded) =
+        buckets.values.where((e) => e.priority >= priority).fold(
+      (0, 0),
+      (prev, entry) {
+        final downloaded = entry.sinceLast;
+        final total = entry.targetCount - entry.atLast;
+        return (prev.$1 + total, prev.$2 + downloaded);
+      },
+    );
+
+    return ProgressWithOperations._(total, downloaded);
+  }
+
+  InternalSyncDownloadProgress incrementDownloaded(SyncDataBatch batch) {
+    final newBucketStates = Map.of(buckets);
+    for (final dataForBucket in batch.buckets) {
+      final previous = newBucketStates[dataForBucket.bucket]!;
+      newBucketStates[dataForBucket.bucket] = (
+        priority: previous.priority,
+        atLast: previous.atLast,
+        sinceLast: previous.sinceLast + dataForBucket.data.length,
+        targetCount: previous.targetCount,
+      );
+    }
+
+    return InternalSyncDownloadProgress(newBucketStates);
+  }
+
+  SyncDownloadProgress get asSyncDownloadProgress =>
+      SyncDownloadProgress._(this);
+
+  @override
+  int get hashCode => _mapEquality.hash(buckets);
+
+  @override
+  bool operator ==(Object other) {
+    return other is InternalSyncDownloadProgress &&
+        // totalOperations and downloadedOperations are derived values, but
+        // comparing them first helps find a difference faster.
+        totalOperations == other.totalOperations &&
+        downloadedOperations == other.downloadedOperations &&
+        _mapEquality.equals(buckets, other.buckets);
+  }
+
+  @override
+  String toString() {
+    final all = asSyncDownloadProgress;
+    return 'for total: ${all.downloadedOperations} / ${all.totalOperations}';
+  }
+
+  static const _mapEquality = MapEquality<Object?, Object?>();
+}
+
+/// Information about a progressing download.
+///
+/// This reports the `total` amount of operations to download, how many of them
+/// have alreaady been `completed` and finally a `fraction` indicating relative
+/// progress (as a number between `0.0` and `1.0`)
+///
+/// To obtain these values, use [SyncDownloadProgress] available through
+/// [SyncStatus.downloadProgress].
+final class ProgressWithOperations {
+  /// How many operations need to be downloaded in total until the current
+  /// download is complete.
+  final int totalOperations;
+
+  /// How many operations have already been downloaded since the last complete
+  /// download.
+  final int downloadedOperations;
+
+  ProgressWithOperations._(this.totalOperations, this.downloadedOperations);
+
+  /// Relative progress (as a number between `0.0` and `1.0`).
```

Comment:
Maybe explicitly note there could be a progress of 1.0 even though we're busy saving the data.